### DISCUSSION TITLE
Add self-referential nodes to egraph

### DISCUSSION THREAD

**Discussion by silversquirl** [egraphs-good/egg#240] [Q&A] [answered]
Is it possible to add an ENode that references itself? Or, more generally, to add a cycle to the EGraph?
It seems like it must be doable since egg itself creates self-references using rewrites, but I'm not sure how to do it with the user-facing API.

**Comment by mwillsey** [comment, marked as answer]
Yes! You cannot directly make a cycle, but you can do it with a union. Here is some (un-parsed) code to demonstrate:

```rust
let x = egraph.add(Leaf("x"));
let node = egraph.add(Node(x, x));
egraph.union(x, node);
```
